dashboard/config: refactor config system

dvyukov opened the pull request:
*******************************************************************************
Before sending a pull request, please review Contribution Guidelines:
https://github.com/google/syzkaller/blob/master/docs/contributing.md
*******************************************************************************

melver commented on dashboard/config/linux/bits/base.yml:
```diff
@@ -0,0 +1,124 @@
+ # Common config fragments required by syzbot for all kernels
+
+verbatim:
+ # This config can be used to enable any additional temporal debugging features in linux-next tree.
```
I think this was also copied into the final .config.

If so, it needs to be an actual verbatim block of text using `verbatim: |`

dvyukov replied:
I will check what this option does to newlines. I saw most of these multi-line options mess with newlines too much.

dvyukov replied:
Done.

blackgnezdo commented on pkg/vcs/openbsd.go:
```diff
@@ -7,8 +7,8 @@ type openbsd struct {
 }
 
-func newOpenBSD(vm, dir string) *openbsd {
+func newOpenBSD(dir string, user bool) *openbsd {
```
`bool` seem like a fairly suboptimal type for something named `user`. Can this be an enum?

dvyukov replied:
Done

blackgnezdo commented on pkg/vcs/vcs.go:
```diff
@@ -151,28 +155,42 @@ type BisectEnv struct {
+}
+
+// NewUserRepo for command-line tools that work with a user-provided repo.
+// Such repo will not try to srop and re-create the checkout, but rather return errors,
```
drop, not srop

dvyukov replied:
Done

blackgnezdo commented on pkg/vcs/vcs.go:
```diff
@@ -151,28 +155,42 @@ type BisectEnv struct {
 }
 
+// NewRepo is intended for autonomous programs that fully manage the repo.
+// Such repo will try to re-create the checkout from scratch to recover from any errors,
```
Looks like what you are after is "precious" vs. "discardable" repo, rather than ["not"] "user". Maybe adapt the terminology accordingly. Also would work better with enum.

dvyukov replied:
Renamed to precious and added typed options.

blackgnezdo commented on tools/syz-kconf/parser.go:
```diff
@@ -0,0 +1,284 @@
+// Copyright 2020 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package main
```
Maybe some comment about what this does? Or is it documented in a .md somewhere?

dvyukov replied:
There is README.md with more details here:
https://github.com/google/syzkaller/pull/1975/commits/d5b2aa4a509ffb7cded9d63100216eb3ead3a201

I've added a short package comment to point to that README.md.

dvyukov commented on tools/syz-kconf/kconf.go:
```diff
@@ -0,0 +1,377 @@
+	// Some USB drivers don't depend on any USB related symbols, but rather on a generic symbol
+	// for some input subsystem (e.g. HID), so include it as well.
+	// TODO: don't include HID, it does not seem to add anything relevant.
+	return ctx.addDependentConfigs(cf, []string{"USB_SUPPORT", "HID"}, []string{"USB_GADGET"}, distroConfig)
```
@xairy PTAL

You script also use USB_ARCH_HAS_HCD and USB. But these already depend on USB_SUPPORT, so don't seem to be necessary. Maybe the python library mishandled transitive dependencies?

I am also not sure why HID is need. If I remove it, I see only some irrelevant configs go away, e.g. ps/2 mouse. I will remove it in a separate commit.

Re these configs:
https://github.com/google/syzkaller/blob/ff4a3345a1b2a40ff1b8b983153d0b1fcc72f1c5/dashboard/config/util.sh#L47-L72
Did you want to remove everything that depends on USB_GADGET?

xairy replied:
> You script also use USB_ARCH_HAS_HCD and USB. But these already depend on USB_SUPPORT, so don't seem to be necessary.

This could be.

> I am also not sure why HID is need. If I remove it, I see only some irrelevant configs go away, e.g. ps/2 mouse. I will remove it in a separate commit.

AFAIR HID uses its own drivers dispatching system, and some of the HID drivers don't depend on USB directly. Maybe something works differently with your parser...

> Did you want to remove everything that depends on USB_GADGET?

No, the intention was to disable all legacy gadget drivers (see `drivers/usb/gadget/legacy/Kconfig`) expect for `USB_RAW_GADGET`.

[...]

Looking at the new USB config: `CONFIG_USB_DUMMY_HCD` and `CONFIG_USB_RAW_GADGET ` must be enabled.

dvyukov replied:
> AFAIR HID uses its own drivers dispatching system, and some of the HID drivers don't depend on USB directly. Maybe something works differently with your parser...

The new parser extracts dependencies transitively, however I think the old one did it as well.

I will prepare a change and we will look at the actual, then it will be easier to decide.
Maybe that's because I disabled everything that depends on USB_GADGET, thus I did not see anything useful that depends on HID.

dvyukov replied:
Fixed things.

CONFIG_USB_RAW_GADGET and CONFIG_USB_DUMMY_HCD are now enabled.
Also, we don't disable everything that depends on USB_GADGET, but rather disable the specific configs you disabled before.

However, there is still a significant diff. Some MEDIA/DVB configs are not enabled now, but they don't depend on USB, so I am not sure why they were enabled before...

dvyukov commented:
For upstream-kasan.config I managed to get almost no diff. And it tested both upstream-kasan.config and upstream-kasan-base.config, both work fine.

dvyukov commented:
For KCSAN there are ~1000 lines changed, but it boots/works and the changes look reasonable, so I propose to merge it as is, hopefully in future we will have more control and understanding over this.
I've looked at one USB config -- USB_DWC3_ULPI, it is now disabled because it depends on USB_GADGET. @xairy is it fine change? You seemed to want to disable configs that depend on USB_GADGET.
Some filesystem configs become enabled, but that's good, we just did not sync with KASAN config.
@melver

dvyukov commented:
@xairy If we merge USB configs into upstream-kasan.config, are there any reasons to keep separate upstream-usb.config?
Also, should we enable all USB? Rather then only the ones that are enabled in distros?

melver commented:
> For KCSAN there are ~1000 lines changed, but it boots/works and the changes look reasonable, so I propose to merge it as is, hopefully in future we will have more control and understanding over this.
> I've looked at one USB config -- USB_DWC3_ULPI, it is now disabled because it depends on USB_GADGET. @xairy is it fine change? You seemed to want to disable configs that depend on USB_GADGET.
> Some filesystem configs become enabled, but that's good, we just did not sync with KASAN config.
> @melver

LGTM. I think it looks fine, most changes are only related to specific drivers or subsystems.

xairy commented:
> @xairy If we merge USB configs into upstream-kasan.config, are there any reasons to keep separate upstream-usb.config?

The intention was to not include unnecessary stuff for USB fuzzing config. There was (is?) a lot of configs enabled in upstream-kasan.cfg for which no one knows where those came from. Currently the USB config generation is a fully scripted based on distro configs.

But generally I don't mind merging.

> Also, should we enable all USB? Rather then only the ones that are enabled in distros?

I think we'll be getting way too many bugs that no one cares about this way.

dvyukov commented:
For KMSAN there is also ~1000 lines changed, similar to KCSAN, so not bad in itself.

Some things we can now do even better. E.g. dashboard/config/bits-kmsan.config contains:
```
# CONFIG_PARAVIRT is not set
```
but upstream-kmsan.config actually contains:
```
CONFIG_PARAVIRT=y
```
Now we disable it for real. Though, since current config works, perhaps we don't need to disable it anymore...

For both new and old KMSAN configs I see the following warning when booting kernel.
The ioremap error went away, I don't an explanation for it...

```
[   18.134145][    T1] ------------[ cut here ]------------
[   18.135669][    T1] WARNING: CPU: 3 PID: 1 at drivers/gpu/drm/drm_gem_vram_helper.c:279 drm_gem_vram_offset+0x168/0x180
[   18.135688][    T1] Kernel panic - not syncing: panic_on_warn set ...
[   18.135718][    T1] CPU: 3 PID: 1 Comm: swapper/0 Not tainted 5.9.0-rc8+ #52
[   18.135733][    T1] Hardware name: QEMU Standard PC (i440FX + PIIX, 1996), BIOS rel-1.13.0-44-g88ab0c15525c-prebuilt.qemu.org 04/01/2014
[   18.135746][    T1] Call Trace:
[   18.135782][    T1]  dump_stack+0x21c/0x280
[   18.135807][    T1]  panic+0x4c8/0xea7
[   18.135848][    T1]  ? kmsan_get_shadow_origin_ptr+0x81/0xb0
[   18.135868][    T1]  __warn+0x416/0x6a0
[   18.135898][    T1]  ? drm_gem_vram_offset+0x168/0x180
[   18.135921][    T1]  report_bug+0x7d4/0x9d0
[   18.135948][    T1]  ? drm_gem_vram_offset+0x168/0x180
[   18.135980][    T1]  handle_bug+0x6c/0xd0
[   18.136006][    T1]  __exc_invalid_op+0x34/0x80
[   18.136032][    T1]  exc_invalid_op+0x30/0x40
[   18.136056][    T1]  asm_exc_invalid_op+0x12/0x20
[   18.136088][    T1] RIP: 0010:drm_gem_vram_offset+0x168/0x180
[   18.136120][    T1] Code: ff ff 31 c0 31 c9 eb ae 8b 7d d4 e8 b2 c9 73 fb e9 4b ff ff ff 8b 3a e8 a6 c9 73 fb 4d 85 ff 0f 85 63 ff ff ff e8 a8 8e f7 fa <0f> 0b 31 c0f
[   18.136133][    T1] RSP: 0000:ffff88813d7326a8 EFLAGS: 00010293
[   18.136157][    T1] RAX: ffffffff86bd55b8 RBX: ffff888133d7a580 RCX: ffff88823da90000
[   18.136172][    T1] RDX: 0000000000000000 RSI: 00000000000004c0 RDI: ffff888133d7a580
[   18.136190][    T1] RBP: ffff88813d7326e0 R08: ffffea000000000f R09: ffff88813fffa000
[   18.136205][    T1] R10: 0000000000000004 R11: 00000000ffffffff R12: ffff888133d7a400
[   18.136220][    T1] R13: 0000000000000000 R14: ffff88823da909e0 R15: 0000000000000000
[   18.136247][    T1]  ? drm_gem_vram_offset+0x168/0x180
[   18.136274][    T1]  ? drm_gem_vram_offset+0x168/0x180
[   18.136300][    T1]  bochs_pipe_enable+0x1a4/0x480
[   18.136333][    T1]  drm_simple_kms_crtc_enable+0x138/0x1b0
[   18.136356][    T1]  ? bochs_connector_get_modes+0x210/0x210
[   18.136382][    T1]  ? drm_simple_kms_crtc_check+0x280/0x280
[   18.136409][    T1]  drm_atomic_helper_commit_modeset_enables+0x41a/0x12b0
[   18.136440][    T1]  drm_atomic_helper_commit_tail+0xe2/0x960
[   18.136468][    T1]  ? kmsan_get_metadata+0x116/0x180
[   18.136492][    T1]  commit_tail+0x7c9/0x9f0
[   18.136523][    T1]  drm_atomic_helper_commit+0xecc/0xf60
[   18.136551][    T1]  ? drm_atomic_commit+0xa7/0x220
[   18.136585][    T1]  ? drm_atomic_helper_async_commit+0x900/0x900
[   18.136609][    T1]  drm_atomic_commit+0x1c3/0x220
[   18.136643][    T1]  drm_client_modeset_commit_atomic+0xac9/0xd60
[   18.136681][    T1]  drm_client_modeset_commit_locked+0x16b/0x990
[   18.136711][    T1]  ? __msan_metadata_ptr_for_load_8+0x10/0x20
[   18.136740][    T1]  drm_client_modeset_commit+0x9e/0x120
[   18.136774][    T1]  __drm_fb_helper_restore_fbdev_mode_unlocked+0x200/0x350
[   18.136800][    T1]  ? kmsan_get_metadata+0x80/0x180
[   18.136830][    T1]  drm_fb_helper_set_par+0x1b6/0x2a0
[   18.136862][    T1]  ? drm_fb_helper_fill_pixel_fmt+0x7f0/0x7f0
[   18.136891][    T1]  fbcon_init+0x1de0/0x2e80
[   18.136921][    T1]  ? process_slab+0x14d0/0x18a0
[   18.136951][    T1]  ? fbcon_startup+0x1780/0x1780
[   18.136975][    T1]  visual_init+0x412/0x8b0
[   18.137004][    T1]  do_bind_con_driver+0x129c/0x1c40
[   18.137039][    T1]  do_take_over_console+0xf41/0x10e0
[   18.137071][    T1]  ? kmsan_get_shadow_origin_ptr+0x81/0xb0
[   18.137102][    T1]  fbcon_fb_registered+0xa05/0xdd0
[   18.137131][    T1]  register_framebuffer+0xe96/0x10d0
[   18.137169][    T1]  __drm_fb_helper_initial_config_and_unlock+0x1e71/0x26d0
[   18.137204][    T1]  drm_fbdev_client_hotplug+0xb88/0xf40
[   18.137229][    T1]  drm_fbdev_generic_setup+0x39d/0xa00
[   18.137253][    T1]  bochs_pci_probe+0x8ab/0x8c0
[   18.137278][    T1]  ? qxl_gem_prime_mmap+0x30/0x30
[   18.137300][    T1]  pci_device_probe+0xae6/0xe80
[   18.137325][    T1]  ? pci_uevent+0x820/0x820
[   18.137353][    T1]  really_probe+0xdd2/0x2420
[   18.137386][    T1]  driver_probe_device+0x293/0x390
[   18.137416][    T1]  device_driver_attach+0x25a/0x490
[   18.137445][    T1]  __driver_attach+0x78c/0x840
[   18.137471][    T1]  ? kmsan_get_metadata+0x116/0x180
[   18.137498][    T1]  bus_for_each_dev+0x210/0x340
[   18.137523][    T1]  ? driver_attach+0xb0/0xb0
[   18.137551][    T1]  driver_attach+0x89/0xb0
[   18.137577][    T1]  bus_add_driver+0x688/0xc70
[   18.137602][    T1]  driver_register+0x485/0x8e0
[   18.137627][    T1]  __pci_register_driver+0x216/0x370
[   18.137658][    T1]  bochs_init+0x135/0x17d
[   18.137684][    T1]  do_one_initcall+0x371/0x9c0
[   18.137706][    T1]  ? qxl_init+0x1d3/0x1d3
[   18.137741][    T1]  ? kmsan_get_metadata+0x116/0x180
[   18.137769][    T1]  ? kmsan_get_shadow_origin_ptr+0x81/0xb0
[   18.137789][    T1]  ? qxl_init+0x1d3/0x1d3
[   18.137817][    T1]  do_initcall_level+0x1e5/0x3c6
[   18.137844][    T1]  do_initcalls+0x127/0x1cb
[   18.137870][    T1]  ? cpu_init_udelay+0xcf/0xcf
[   18.137893][    T1]  do_basic_setup+0x33/0x36
[   18.137918][    T1]  kernel_init_freeable+0x23d/0x390
[   18.137941][    T1]  ? rest_init+0x1f0/0x1f0
[   18.137960][    T1]  kernel_init+0x1f/0x840
[   18.137979][    T1]  ? rest_init+0x1f0/0x1f0
[   18.138001][    T1]  ret_from_fork+0x1f/0x30
[   18.138349][    T1] Kernel Offset: disabled
[   18.144021][    T1] Rebooting in 86400 seconds..
```

@ramosian-glider

dvyukov commented:
> > @xairy If we merge USB configs into upstream-kasan.config, are there any reasons to keep separate upstream-usb.config?
> 
> The intention was to not include unnecessary stuff for USB fuzzing config. There was (is?) a lot of configs enabled in upstream-kasan.cfg for which no one knows where those came from. Currently the USB config generation is a fully scripted based on distro configs.
> 
> But generally I don't mind merging.
>
> > Also, should we enable all USB? Rather then only the ones that are enabled in distros?
> 
> I think we'll be getting way too many bugs that no one cares about this way.

OK, let's leave everything as it is now for now.

dvyukov commented:
@xairy please take a look at:
https://github.com/google/syzkaller/pull/1975/commits/56aea639afd7306370514c61ab4ef836f5dd830e
It removes HID. I don't see anything obviously useful among disabled things. Some mouses and LEDs. At least no config has USB in the name.

xairy commented:
> @xairy please take a look at:
> [56aea63](https://github.com/google/syzkaller/commit/56aea639afd7306370514c61ab4ef836f5dd830e)
> It removes HID. I don't see anything obviously useful among disabled things. Some mouses and LEDs. At least no config has USB in the name.

This does not look good. It's the thing I mentioned, HID configs don't depend on USB directly.

E.g. this change disables `CONFIG_HID_PRODIKEYS`, which is reachable over USB, and we've actually found [a bug](https://syzkaller.appspot.com/bug?id=7abf8bebe4daf009ccd1c65a408f63d94583438b) in it.

xairy commented:
Would it make sense to add tests against the generated configs? E.g. make sure that KASAN config always has KASAN enabled, and USB config always has Raw Gadget and Dummy HCD?

dvyukov commented:
> > @xairy please take a look at:
> > [56aea63](https://github.com/google/syzkaller/commit/56aea639afd7306370514c61ab4ef836f5dd830e)
> > It removes HID. I don't see anything obviously useful among disabled things. Some mouses and LEDs. At least no config has USB in the name.
> 
> This does not look good. It's the thing I mentioned, HID configs don't depend on USB directly.
> 
> E.g. this change disables `CONFIG_HID_PRODIKEYS`, which is reachable over USB, and we've actually found [a bug](https://syzkaller.appspot.com/bug?id=7abf8bebe4daf009ccd1c65a408f63d94583438b) in it.

I see. I dropped this commit.
But I disabled at least HYPERV, which is enabled via HID_HYPERV_MOUSE. And HYPERV brings in lots of other configs.

dvyukov commented:
> e sure that KASAN config always ha

dvyukov commented:
> Would it make sense to add tests against the generated configs? E.g. make sure that KASAN config always has KASAN enabled, and USB config always has Raw Gadget and Dummy HCD and make some selected drivers?

This is one of the main points of this config system. If something is specified and has a wrong value in the resulting config, it produces an error.
The problem is that none of USB configs are spelled out explicitly. But we can spell at least few major ones explicitly.

dvyukov commented:
> > Would it make sense to add tests against the generated configs? E.g. make sure that KASAN config always has KASAN enabled, and USB config always has Raw Gadget and Dummy HCD and make some selected drivers?
> 
> This is one of the main points of this config system. If something is specified and has a wrong value in the resulting config, it produces an error.
> The problem is that none of USB configs are spelled out explicitly. But we can spell at least few major ones explicitly.

Well, we actually now do. In the recent versions we now have in usb.yml:
```
config:
 - USB_SUPPORT
 - USB_GADGET
 - USB_RAW_GADGET: [v5.7]
 - USB_DUMMY_HCD
```
so if any of these are not present, the generation process will fail.

xairy commented:
Found another issue: the new config enables some USB configs that aren't enabled in any of the distro configs, e.g. `CONFIG_RTLWIFI_DEBUG` and probably some others. In the python script we disable those as a separate step. For normal config these flaky ones don't really matter, but that step was specifically added to not enabled anything unneeded for Android fuzzing.

xairy commented:
`CONFIG_HID_REDRAGON` is an example of the same issue, but for a driver config.

dvyukov commented:
> `CONFIG_HID_REDRAGON` is an example of the same issue, but for a driver config.

CONFIG_HID_REDRAGON is provided by defconfig.
Yes, the new code only enables configs.
Question is: are there any defconfig's that are not important? Are there enough to care to disable them?
E.g. CONFIG_HID_REDRAGON is enabled on my desktop and I see it enabled in ClearLinux:
https://github.com/clearlinux-pkgs/kernel-config/blob/master/base-5.4
So perhaps the majority we got disabled this way just point to our incomplete set of distro configs?

xairy commented:
I guess it's OK for upstream-kasan.cfg, but I'd not do that for Android configs.

Looking at android-5.4 config we end up enabling e.g. `CONFIG_HID_STEAM` and `CONFIG_USB_UAS` which aren't enabled by defconfig nor by any of our android-* distro configs.

Also noticed that android-5.4 should have `CONFIG_USB_RAW_GADGET` enabled, but this change disables it.

dvyukov commented:
> Looking at android-5.4 config we end up enabling e.g. CONFIG_HID_STEAM and CONFIG_USB_UAS which aren't enabled by defconfig nor by any of our android-* distro configs.

Both of these come from gki_defconfig.

xairy commented:
Ah, I see. I wonder why does Pixel not have them enabled? OK, enabling stuff that comes from GKI config makes sense.

Alright, except for `CONFIG_USB_RAW_GADGET` missing, the rest looks good.

dvyukov commented:
I've pushed a new version.

> Alright, except for CONFIG_USB_RAW_GADGET missing, the rest looks good.

Found the problem. It's because it was cherry-picked into android-5.4, but is not present in stable-5.4, so I did not get declaration right. Now it is:
```
 # It's introduced in v5.5 or so, but we want it in android-5.4, but it's not present in stable-5.4.
 # A better way to express this would to add "or" operator (v5.5||android-5.4), but for now we have this.
 - USB_RAW_GADGET: [v5.4, -stable-5.4]
```
https://github.com/google/syzkaller/pull/1975/commits/552b65eb5705df4855cf0c5db21cda4a876fd07b


> Found another issue: the new config enables some USB configs that aren't enabled in any of the distro configs, e.g. CONFIG_RTLWIFI_DEBUG and probably some others. In the python script we disable those as a separate step. For normal config these flaky ones don't really matter, but that step was specifically added to not enabled anything unneeded for Android fuzzing.

Fixed. This was die to a more general problem: we run olddefconfig to fix up things, but it may enable some things because they are not present in the config and have default value y. This is what happened with CONFIG_RTLWIFI_DEBUG, but also affected lots of other configs.
This is now fixed here:
https://github.com/google/syzkaller/pull/1975/commits/d4236248dab8d79005f8473e53f1fb01735dbe12
(we effectively do something like "oldnoconfig")

dvyukov commented:
Doh! this now doesn't boot with some issue in USB gadget:

```
WARNING: SOFTIRQ-safe -> SOFTIRQ-unsafe lock order detected
5.9.0+ #54 Not tainted
-----------------------------------------------------
v4l_id/5309 [HC0[0]:SC0[0]:HE0:SE1] is trying to acquire:
ffff888041b38e38 (&trig->leddev_list_lock){.+.?}-{2:2}, at: led_trigger_blink_setup drivers/leds/led-triggers.c:403 [inline]
ffff888041b38e38 (&trig->leddev_list_lock){.+.?}-{2:2}, at: led_trigger_blink_setup drivers/leds/led-triggers.c:392 [inline]
ffff888041b38e38 (&trig->leddev_list_lock){.+.?}-{2:2}, at: led_trigger_blink_oneshot+0x39/0xe0 drivers/leds/led-triggers.c:427

and this task is already holding:
ffff8880131a87e0 (&udc->lock){....}-{2:2}, at: spin_lock include/linux/spinlock.h:354 [inline]
ffff8880131a87e0 (&udc->lock){....}-{2:2}, at: get_gadget_descs+0x183/0x570 drivers/usb/usbip/vudc_sysfs.c:42
which would create a new lock dependency:
 (&udc->lock){....}-{2:2} -> (&trig->leddev_list_lock){.+.?}-{2:2}

but this new dependency connects a SOFTIRQ-irq-safe lock:
 (udc_lock){+.+.}-{3:3}

... which became SOFTIRQ-irq-safe at:
  lock_acquire kernel/locking/lockdep.c:5442 [inline]
  lock_acquire+0x1a8/0x6e0 kernel/locking/lockdep.c:5407
  __mutex_lock_common kernel/locking/mutex.c:956 [inline]
  __mutex_lock+0x156/0x13c0 kernel/locking/mutex.c:1103
  usb_add_gadget+0x21f/0x660 drivers/usb/gadget/udc/core.c:1228
  usb_add_gadget_udc_release+0x29/0x70 drivers/usb/gadget/udc/core.c:1281
  dummy_udc_probe+0x6a0/0xa00 drivers/usb/gadget/udc/dummy_hcd.c:1076
  platform_drv_probe+0xce/0x1a0 drivers/base/platform.c:761
  really_probe+0x290/0xc10 drivers/base/dd.c:554
  driver_probe_device+0x26b/0x3d0 drivers/base/dd.c:738
  __device_attach_driver+0x1d1/0x290 drivers/base/dd.c:844
  bus_for_each_drv+0x15f/0x1e0 drivers/base/bus.c:431
  __device_attach+0x26d/0x3f0 drivers/base/dd.c:912
  bus_probe_device+0x1e4/0x290 drivers/base/bus.c:491
  device_add+0xbd6/0x1af0 drivers/base/core.c:2938
  platform_device_add+0x35c/0x820 drivers/base/platform.c:611
  instrument_atomic_read include/linux/instrumented.h:71 [inline]
  test_bit include/asm-generic/bitops/instrumented-non-atomic.h:134 [inline]
  init+0x7cf/0xb93 arch/x86/crypto/camellia_aesni_avx_glue.c:279
  do_one_initcall+0x103/0x650 init/main.c:1207
  do_initcall_level init/main.c:1280 [inline]
  do_initcalls init/main.c:1296 [inline]
  do_basic_setup init/main.c:1316 [inline]
  kernel_init_freeable+0x612/0x696 init/main.c:1516
  kernel_init+0xd/0x1b8 init/main.c:1405
  ret_from_fork+0x1f/0x30 arch/x86/entry/entry_64.S:296
```
https://gist.githubusercontent.com/dvyukov/a11ea3f493da48bc4405dcb73209d424/raw/ee16ebfaab3157fd6f4f874613fd4d775618c42c/gistfile1.txt

dvyukov commented:
Seems to be due to USB_LED_TRIG.
We did not have it in upstream-kasan.config, but we had it in upstream-usb.config, so I am not sure why we did not see it before...
@xairy maybe you can figure out something from the stack trace? What other config may lead to this error?

xairy commented:
Some part of this helps with that:

```
  scripts/config -d CONFIG_USB_G_NCM
  scripts/config -d CONFIG_USB_G_SERIAL
  scripts/config -d CONFIG_USB_G_PRINTER
  scripts/config -d CONFIG_USB_G_NOKIA
  scripts/config -d CONFIG_USB_G_ACM_MS
  scripts/config -d CONFIG_USB_G_MULTI
  scripts/config -d CONFIG_USB_G_HID
  scripts/config -d CONFIG_USB_G_DBGP
  scripts/config -d CONFIG_USB_G_WEBCAM

  scripts/config -d CONFIG_USB_ZERO
  scripts/config -d CONFIG_USB_AUDIO
  scripts/config -d CONFIG_USB_ETH
  scripts/config -d CONFIG_USB_FUNCTIONFS
  scripts/config -d CONFIG_USB_MASS_STORAGE
  scripts/config -d CONFIG_USB_GADGET_TARGET
  scripts/config -d CONFIG_USB_MIDI_GADGET
  scripts/config -d CONFIG_USB_CDC_COMPOSITE

  scripts/config -d CONFIG_USB_GADGETFS
  scripts/config -d CONFIG_USB_LIBCOMPOSITE
  scripts/config -d CONFIG_USB_CONFIGFS

  scripts/config -e CONFIG_USB_GADGET
  scripts/config -e CONFIG_USB_DUMMY_HCD
  scripts/config -e CONFIG_USB_RAW_GADGET
```

Could you also check that USB runtests still work? AFAIR we also needs some of these, so Raw Gadget actually becomes usable, otherwise we default to one of the other legacy gadgets.

dvyukov commented:
Seems to help. Disabled these gadgets on all instances:
https://github.com/google/syzkaller/pull/1975/commits/23e9f109273e9e1b369635bc6d6cf7f78810fddc
Now for more testing.

dvyukov commented:
> Could you also check that USB runtests still work? AFAIR we also needs some of these, so Raw Gadget actually becomes usable, otherwise we default to one of the other legacy gadgets.

vusb_ath9k were failing, I noticed we never explicitly enabled ATH9K configs and they are enabled accidentally on some instances and inconsistent across instances. I've enabled them consistently:
https://github.com/google/syzkaller/pull/1975/commits/60c3410b5df45b8c565822b00f3a67e90ae3308b

But it did not help. These 2 fail with the new kasan and usb configs:
```
vusb_ath9k none/thr                   : FAIL: run 0: call 0 is not executed
vusb_ath9k none/thr/cover             : FAIL: run 0: call 0 is not executed
vusb_ath9k none/thr C                 : FAIL: run 0: call 0 is not finished

vusb_hid namespace/thr                : FAIL: run 0: call 4 is not executed
vusb_hid namespace/thr/cover          : FAIL: run 0: call 4 is not executed
vusb_hid namespace/thr C              : FAIL: run 0: call 4 is not finished
```

But I tested with the old usb config and they fail in the same way. So I consider this is not a regression.

dvyukov commented:
@hogander-unikie FYI is the kernel config overhaul.
In particular this is new config minimization logic:
https://github.com/google/syzkaller/pull/1975/commits/f7a39f95ec63990b47e55d8beb84b16623386101
I am obviously biased, but I can understand it (as compared to config-bisect.pl which I could only almost understand). The actual minimization logic is around 10 lines and is tested. Now we can evolve it to make it smarter. Some direction are outlined in the comment there.
And this change switches pkg/bisect to this new logic:
https://github.com/google/syzkaller/pull/1975/commits/ab73f70c335875ee696ae33e73a3014a026782a4
